sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -346,11 +373,31 @@ class HashJoinIterator(
         // Need to refill the gatherer
         gathererStore.foreach(_.close())
         gathererStore = None
+        var fallbackStreamIter: Option[Iterator[ColumnarBatch]] = None
+        var fallbackBuildIter: Option[Iterator[ColumnarBatch]] = None
         withResource(stream.next()) { cb =>
-          streamTime += (System.nanoTime() - startTime)
-          withResource(GpuProjectExec.project(built.getBatch, boundBuiltKeys)) { builtKeys =>
-            gathererStore = joinGatherer(builtKeys, built, cb)
+          try {
+            streamTime += (System.nanoTime() - startTime)
+            withResource(GpuProjectExec.project(built.getBatch, boundBuiltKeys)) { builtKeys =>
+              gathererStore = joinGatherer(builtKeys, built, cb)
+            }
+          } catch {
+            case _: OutOfMemoryError if fallback.isDefined =>
+              logWarning("Join was too large falling back to CPU")
+              fallbackStreamIter = Some(Seq(GpuColumnVector.incRefCounts(cb)).toIterator ++ stream)
+              fallbackBuildIter = Some(Seq(GpuColumnVector.incRefCounts(built.getBatch)).toIterator)
+              // As soon as we start the fallback the GpuSemaphore could be released so
+              // close everything now just to be sure
+              close()
```

Comment:
At first I thought `close()` should set `fallbackIter = None` when closing and we should be checking for closed first thing in `hasNext`, but now I see that would be very bad given how this works.  It's quite odd to be intentionally using a closed iterator, but I get the motivation.  It would be good to have comments sprinkled in or at least an overview comment on the class describing this unique behavior.

A way to avoid the odd close-myself-then-continue-as-normal behavior is to have this iterator wrap the actual iterator being used.  When we fallback we don't close ourselves but instead close the underlying GPU iterator and fire up the CPU iterator to replace it.  That would also eliminate all the "if fallback" checks sprinkled throughout.  Not must-fix, but maybe a followup if you agree.